R/dryRun.R

```diff
@@ -0,0 +1,175 @@
+#'
+#' * `browser()`, `browseURL()`, and `rstudioapi::askForPassword()` because
+#'    they need an interactive session that your deployment server will lack.
+#'
+#' ## Current limitations
+#'
+#' * `dryRun()` currently offers no way to diagnose problems with
+#'   mismatched R/Python/Quarto/pandoc versions.
+#'
+#' * `dryRun()` doesn't help much with paths. There are two common problems
+#'   it can't help with: using an absolute path and using the wrong case.
+#'   Both of these will work locally but fail on the server. [lint()]
+#'   uses an alternative technique (static analysis) to detect many of these
+#'   cases.
+#'
+#' @inheritParams deployApp
+#' @param contentCategory Set this to `"site"` if you'd deploy with
+#'   [deploySite()]; otherwise leave as is.
+#' @export
+dryRun <- function(appDir = getwd(),
+                   envVars = NULL,
+                   appFiles = NULL,
+                   appFileManifest = NULL,
+                   appPrimaryDoc = NULL,
+                   contentCategory = NULL,
+                   quarto = NA) {
+  check_installed("callr")
+
+  appFiles <- listDeploymentFiles(
+    appDir,
+    appFiles = appFiles,
+    appFileManifest = appFileManifest
+  )
+
+  appMetadata <- appMetadata(
+    appDir = appDir,
+    appFiles = appFiles,
+    appPrimaryDoc = appPrimaryDoc,
+    quarto = quarto,
+    contentCategory = contentCategory,
+  )
+
+  # copy files to bundle dir to stage
+  cli::cli_alert_info("Bundling app")
+  bundleDir <- bundleAppDir(
+    appDir = appDir,
+    appFiles = appFiles,
+    appPrimaryDoc = appMetadata$appPrimaryDoc
+  )
+  defer(unlink(bundleDir, recursive = TRUE))
+
+  cli::cli_alert_info("Creating project specific library")
+  callr::r(
+    function() {
+      options(renv.verbose = FALSE)
+      renv::init()
+      renv::restore()
+    },
+    wd = bundleDir
+  )
+
+  # Add tracing code -------------------------------------------------
+  cli::cli_alert_info("Adding shims")
+
+  file.copy(
+    system.file("dryRunTrace.R", package = "rsconnect"),
+    file.path(bundleDir, "__rsconnect-dryRunTrace.R")
+  )
+  appendLines(
+    file.path(bundleDir, ".Rprofile"),
+    c("", 'source("__rsconnect-dryRunTrace.R")')
+  )
+
+  # Run ---------------------------------------------------------------
+  cli::cli_alert_info("Starting {appMetadata$appMode}")
+  if (appMetadata$appMode %in% c("rmd-shiny", "quarto-shiny", "shiny", "api")) {
+    cli::cli_alert_warning("Terminate the app to complete the dry run")
+  }
+
+  envVarNames <- setdiff(userEnvVars(), c(envVars, "PATH"))
+  envVarReset <- c(rep_named(envVarNames, ""), callr::rcmd_safe_env())
+
+  callr::r(
+    appRunner(appMetadata$appMode),
+    args = list(primaryDoc = appMetadata$appPrimaryDoc),
+    env = envVarReset,
+    wd = bundleDir,
+    show = TRUE
+  )
+  invisible()
+}
+
+appRunner <- function(appMode) {
+  switch(appMode,
+    "rmd-static" = ,
+    "rmd-shiny" = function(primaryDoc) {
```

Comment:
`rmd-shiny` should use `rmarkdown::run(file = NULL, dir = getwd())`

Reply:
I already set the working directory above, but I'll take the other changes.

Reply:
The `::run(..)` signature is necessary to have R Markdown do its "run the whole directory" magic.. it's not setting the working directory. I think that's this logic: https://github.com/rstudio/rmarkdown/blob/main/R/shiny.R#L83-L107

Reply:
I just meant the `dir` argument

Reply:
OK. Once you give `file = NULL`, the `dir` value is necessary, as its default value of `dirname(file)` is meaningless.